transport/cert/default_cert.go

```diff
@@ -106,5 +114,14 @@ func (s *secureConnectSource) getClientCertificate(info *tls.CertificateRequestI
 	if err != nil {
 		return nil, err
 	}
+	defaultSourceCachedCert = &cert
 	return &cert, nil
 }
+
+func isCertificateExpired(cert *tls.Certificate) bool {
+	parsed, err := x509.ParseCertificate(cert.Certificate[0])
```

Comment:
Does there need to be a bounds check here?

Reply:
Is there a possible race condition here? We are now storing and returning a some internal state that is guarded by a mutex internally, but not externally. References could change Certificates, right?

Reply:
I will add a bound check for cert.Certificate (this function is private, and I'll add a comment to say we expect the parameter to always be a non-nil cert with 1 Certificate).
Race condition shouldn't be possible on the cached cert since getClientCertificate is the only function that reads/writes the cached cert, and everything inside this function including the call to check isCertificateExpired is serialized. The returned tls.Certificate should be read-only by design, though I'm not sure how to enforce this in Golang. Let me know if you have any suggestions, or a specific counter-example in mind to illustrate your concern here. Thanks!